### DISCUSSION TITLE
Virus on network with sigma.js library

### DISCUSSION THREAD

**Discussion by LimeEric** [mesa/mesa#1292] [Q&A]
Hi. I was trying to change the "d3" library, that is used to render the visualization in the virus_on_network example, to "sigma". So I edited the following line in server.py:
`network = NetworkModule(network_portrayal, 500, 500, library="sigma")`

However, sigma requires IDs for nodes and edges. I added IDs based on boltzmann_wealth_model_network (since it uses sigma) in server.py of virus_on_network:

**def network_portrayal(G):**
    
    def node_color(agent):
        ...
    def edge_color(agent1, agent2):
        ...
    def edge_width(agent1, agent2):
        ...
    def get_agents(source, target):
        ...

    portrayal = dict()
    portrayal["nodes"] = [
        {
            "id": node_id,    #added this id
            "size": 6,
            "color": node_color(agents[0]),
            "tooltip": f"id: {agents[0].unique_id}<br>state: {agents[0].state.name}",
        }
        for (node_id, agents) in G.nodes.data("agent")    #changed this line
    ]

    portrayal["edges"] = [
        {
            "edge_id": edge_id,    #added this id
            "source": source,
            "target": target,
            "color": edge_color(*get_agents(source, target)),
            "width": edge_width(*get_agents(source, target)),
        }
        for edge_id, (source, target) in enumerate(G.edges)     #changed this line
    ]
    return portrayal

The server runs without any errors or warnings in the terminal but the network doesn't show, while anything else is visible.
![image](https://user-images.githubusercontent.com/38434664/165118523-657fc212-bc72-487d-9484-f7fa4961f360.png)

The DevTools console shows the following error:

> Uncaught The edge must have a string or number id.    sigma.min.js:2

I appreciate any help because I cannot figure out what I did wrong.

**Comment by rht** [comment]
The Sigma.js implementation was removed in current main branch, because the D3.js does everything Sigma.js, and can do more. See https://github.com/projectmesa/mesa/commit/e5491079e48c95693af20afa81289a24b8bd3138.

**Reply by rht** [reply]
@LimeEric  wait, reading the error message, it seems that your sigma.min.js is still available. This means that you're using the latest release version, not Git. Anyway, may I know why you are using the Sigma.js?

**Reply by LimeEric** [reply]
I wanted to change the library to test the performance since on https://www.sigmajs.org/ it said it's more suitable for larger graphs. The goal would be to have something similar to virus_on_network but with a few 100 or maybe 1000 of nodes and connections. At the moment, I don't think d3 or sigma is the reason for the bad performance with that many nodes, but maybe the small web server?

I haven't done much research on it yet, sorry. Maybe you have some tips regarding the performance?

**Reply by rht** [reply]
Have you tried running the model without the visualization, and observed that it is still slow? If so, then the visualization is not the cause of the slowness.
